extensions/control-plane/provision/provision-bigquery/src/test/java/org/eclipse/edc/connector/provision/gcp/BigQueryProvisionerTest.java

```diff
@@ -0,0 +1,239 @@
+        var bigQueryProvisioner = BigQueryProvisioner.Builder.newInstance(gcpConfiguration, null, null, monitor)
+                .build();
+
+        assertThat(bigQueryProvisioner.canProvision(new ResourceDefinition() {
+            @Override
+            public <RD extends ResourceDefinition, B extends Builder<RD, B>> B toBuilder() {
+                return null;
+            }
+        })).isFalse();
+    }
+
+    @Test
+    void testCanDeprovisionTrue() {
+        var bigQueryProvisioner = BigQueryProvisioner.Builder.newInstance(gcpConfiguration, null, null, monitor)
+                .build();
+
+        var provisionedResource = BigQueryProvisionedResource.Builder.newInstance()
+                .id(RESOURCE_ID)
+                .transferProcessId(TRANSFER_ID)
+                .resourceDefinitionId(RESOURCE_DEFINITION_ID)
+                .resourceName(RESOURCE_NAME)
+                .build();
+
+        assertThat(bigQueryProvisioner.canDeprovision(provisionedResource)).isTrue();
+    }
+
+    @Test
+    void testCanDeprovisionFalse() {
+        var bigQueryProvisioner = BigQueryProvisioner.Builder.newInstance(gcpConfiguration, null, null, monitor)
+                .build();
+
+        assertThat(bigQueryProvisioner.canDeprovision(new ProvisionedResource() {
+        })).isFalse();
+    }
+
+    @Test
+    void provisionSuccessUsingAdc() {
+        provisionSucceeds(null);
+    }
+
+    @Test
+    void provisionSuccessUsingServiceAccount() {
+        provisionSucceeds(TEST_SERVICE_ACCOUNT_NAME);
+    }
+
+    @Test
+    void provisionFailsIfTableDoesntExist() {
+        var bqFactory = new BigQueryFactoryTest();
+        var bigQueryProvisioner = BigQueryProvisioner.Builder.newInstance(gcpConfiguration, bqFactory, null, monitor)
+                .build();
+
+        var resourceDefinition = BigQueryResourceDefinition.Builder.newInstance()
+                .id(RESOURCE_ID)
+                .transferProcessId(TRANSFER_ID)
+                .property(BigQueryService.PROJECT, TEST_PROJECT)
+                .property(BigQueryService.DATASET, TEST_DATASET)
+                .property(BigQueryService.TABLE, TEST_TABLE)
+                .property(BigQueryService.CUSTOMER_NAME, CUSTOMER_NAME)
+                .build();
+
+        var policy = Policy.Builder.newInstance().build();
+
+        var table = mock(Table.class);
+        when(table.exists()).thenReturn(false);
+        when(bigQuery.getTable(TEST_TARGET.getTableId())).thenReturn(table);
+
+        var result = bigQueryProvisioner.provision(resourceDefinition, policy);
+        assertThat(result).succeedsWithin(1, SECONDS)
+                .extracting(StatusResult::failed)
+                .isEqualTo(true);
+    }
+
+    private class BigQueryFactoryTest implements BigQueryFactory {
+        public BigQuery createBigQuery(String serviceAccountName) throws IOException {
+            return bigQuery;
+        }
+    }
+
+    private void provisionSucceeds(String serviceAccountName) {
```

Comment:
this is not a safe way to test, because you pretty much replicated the production code logic, so it's more error prone, please structure the tests in the way that they are understandable from the test method itself.

Reply:
Really good point, as exchanged, I make a first try, by clearly showing the setup using default credentials, and the setup for testing the use of specific service account